event/datadeal/data_delivery_event.go

```diff
@@ -0,0 +1,235 @@
+package datadeal
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	datadealtypes "github.com/medibloc/panacea-core/v2/x/datadeal/types"
+	oracletypes "github.com/medibloc/panacea-core/v2/x/oracle/types"
+	"github.com/medibloc/panacea-doracle/config"
+	"github.com/medibloc/panacea-doracle/crypto"
+	"github.com/medibloc/panacea-doracle/event"
+	"github.com/medibloc/panacea-doracle/panacea"
+	log "github.com/sirupsen/logrus"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+)
+
+type DataDeliveryVoteEvent struct {
+	reactor event.Reactor
+}
+
+func NewDataDeliveryVoteEvent(s event.Reactor) DataDeliveryVoteEvent {
+	return DataDeliveryVoteEvent{s}
+}
+
+func (e DataDeliveryVoteEvent) GetEventType() string {
+	return "data_delivery"
+}
+
+func (e DataDeliveryVoteEvent) GetEventAttributeKey() string {
+	return "vote_status"
+}
+
+func (e DataDeliveryVoteEvent) GetEventAttributeValue() string {
+	return "'started'"
+}
+
+func (e DataDeliveryVoteEvent) EventHandler(event ctypes.ResultEvent) error {
+
+	dealIDStr := event.Events[datadealtypes.EventTypeDataDeliveryVote+"."+datadealtypes.AttributeKeyDealID][0]
+	dataHash := event.Events[datadealtypes.EventTypeDataDeliveryVote+"."+datadealtypes.AttributeKeyDataHash][0]
+
+	dealID, err := strconv.ParseUint(dealIDStr, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	oraclePrivKey := e.reactor.OraclePrivKey()
+
+	voteOption, deliveredCid, err := e.verifyAndGetVoteOption(dealID, dataHash, oraclePrivKey)
+	if err != nil {
+		log.Infof("error while verify: %v", err)
+		if voteOption == oracletypes.VOTE_OPTION_UNSPECIFIED {
+			return err
+		}
+	}
+
+	msgVoteDataDelivery, err := e.makeDataDeliveryVote(
+		e.reactor.OracleAcc().GetAddress(),
+		dataHash,
+		deliveredCid,
+		dealID,
+		voteOption,
+		oraclePrivKey.Serialize(),
+	)
+	if err != nil {
+		return err
+	}
+
+	txBuilder := panacea.NewTxBuilder(*e.reactor.QueryClient())
+
+	txBytes, err := e.generateTxBytes(msgVoteDataDelivery, e.reactor.OracleAcc().GetPrivKey(), e.reactor.Config(), txBuilder)
+	if err != nil {
+		return err
+	}
+
+	if err := e.broadcastTx(e.reactor.GRPCClient(), txBytes); err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func (e DataDeliveryVoteEvent) verifyAndGetVoteOption(dealID uint64, dataHash string, oraclePrivKey *btcec.PrivateKey) (oracletypes.VoteOption, string, error) {
```

Comment:
I think it would be good write log when every case return from this function.
Because operators may want to know why the vote was decided.

Reply:
Thanks, I fixed it. Please check again :)